cloud/docker/docker.py

```diff
@@ -796,13 +804,22 @@ def get_differing_containers(self):
                     continue
 
             # EXPOSED PORTS
-            # Note that ports that are bound at container run are also exposed
-            # implicitly.
-            expected_exposed_ports = set()
+
+            image_exposed_ports = set((image["ContainerConfig"]["ExposedPorts"] or {}).keys())
+
+            expected_exposed_ports = set(image_exposed_ports)
             for p in (self.exposed_ports or []):
                 expected_exposed_ports.add("/".join(p))
 
-            actually_exposed_ports = set((container["Config"]["ExposedPorts"] or {}).keys())
+            if self.port_bindings:
+                for container_port in self.port_bindings.keys():
+                    if isinstance(container_port, int):
+                        container_port = "{}/tcp".format(container_port)
+
+                    expected_exposed_ports.add(container_port)
+
+            container_exposed_ports = set((container["NetworkSettings"]["Ports"] or {}).keys())
+            actually_exposed_ports = image_exposed_ports.union(container_exposed_ports)
```

Comment:
Also, now that I think about it, it's strange to union `image_exposed_ports` into both `expected_exposed_ports` *and* `actually_exposed_ports`. Doesn't that make it not do anything? You should only need to include it in `expected_exposed_ports`, I would think; they should all be in `actually_exposed_ports` anyway. If they aren't, there's something wrong and we probably need to reboot anyway :wink: